dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -299,37 +305,102 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
         co_await winrt::resume_background();
 
         std::wstring appId{ RetrieveNotificationAppId() };
+
         THROW_IF_FAILED(ToastNotifications_RemoveAllToastsForApp(appId.c_str()));
     }
 
+#if 1 //  Option 1
     winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification>> AppNotificationManager::GetAllAsync()
     {
-        co_await winrt::resume_background();
+        HRESULT telemetryResult{ S_OK };
+        std::exception_ptr eptr{};
+        std::wstring appId{};
 
-        std::wstring appId{ RetrieveNotificationAppId() };
-        winrt::com_ptr<ToastABI::IVector<ToastABI::INotificationProperties*>> toastPropertiesCollection{};
-        auto result{ ToastNotifications_GetHistory(appId.c_str(), toastPropertiesCollection.put()) };
+        auto logTelemetry{ wil::scope_exit([&](){
+            AppNotificationTelemetry::GetAllAsyncByAPI(telemetryResult, appId);
+            RethrowIf(eptr);
+        }) };
+
+        try
+        {
+            co_await winrt::resume_background();
+
+            appId = RetrieveNotificationAppId();
+
+            winrt::com_ptr<ToastABI::IVector<ToastABI::INotificationProperties*>> toastPropertiesCollection{};
+            auto result{ ToastNotifications_GetHistory(appId.c_str(), toastPropertiesCollection.put()) };
+
+            THROW_HR_IF(result, result != S_OK && result != E_NOT_SET); // Swallow E_NOT_SET and return an empty properties vector to signal that there are no active toasts
+
+            unsigned int count{};
+            if (toastPropertiesCollection)
+            {
+                THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+            }
+
+            winrt::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification> toastNotifications{ winrt::single_threaded_vector<winrt::Microsoft::Windows::AppNotifications::AppNotification>() };
+
+            for (unsigned i = 0; i < count; ++i)
+            {
+                ToastABI::INotificationProperties* toastProperties;
+                THROW_IF_FAILED(toastPropertiesCollection->GetAt(i, &toastProperties));
 
-        THROW_HR_IF(result, result != S_OK && result != E_NOT_SET); // Swallow E_NOT_SET and return an empty properties vector to signal that there are no active toasts
+                auto toastNotification{ ToastNotificationFromToastProperties(toastProperties) };
 
-        unsigned int count{};
-        if (toastPropertiesCollection)
+                toastNotifications.Append(toastNotification);
+            }
+
+            co_return toastNotifications;
+        }
+        catch (...)
         {
-            THROW_IF_FAILED(toastPropertiesCollection->get_Size(&count));
+            eptr = std::current_exception();
+            telemetryResult = wil::ResultFromCaughtException();
```

Comment:
This is the reason I prefer Option 2 over Option 1. Capturing the exception in the catch block so it can be rethrown later outside the catch block makes the code more complex to maintain and understand. I much prefer the Option 1 approach which lets the language handle rethrowing the exception from the catch block.

Reply:
We're going with the first option.